File: backend/app/models/review.py
import enum

from sqlalchemy import (
    Boolean, Column, DateTime, ForeignKey, Integer, String, Text,
    UniqueConstraint, func,
)
from sqlalchemy.orm import relationship

from ..database import Base


class ReviewTargetType(str, enum.Enum):
    course  = "course"
    mentor  = "mentor"
    post    = "post"
    event   = "event"


class ReviewStatus(str, enum.Enum):
    pending  = "pending"    # awaiting mediator approval
    approved = "approved"
    rejected = "rejected"


class Review(Base):
    """
    User-written star rating + text for a course, mentor, post, or event.
    One review per (user, target_type, target_id). Requires mediator approval
    before it appears publicly.
    """

    __tablename__ = "reviews"

    id               = Column(Integer, primary_key=True, index=True)
    user_id          = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    target_type      = Column(String, nullable=False)    # ReviewTargetType value
    target_id        = Column(Integer, nullable=False)
    rating           = Column(Integer, nullable=False)   # 1–5
    title            = Column(String(200), nullable=True)
    body             = Column(Text, nullable=False)
    status           = Column(String, default=ReviewStatus.pending.value, nullable=False)
    rejection_reason = Column(Text, nullable=True)
    reviewed_by      = Column(Integer, ForeignKey("users.id"), nullable=True)
    reviewed_at      = Column(DateTime, nullable=True)
    is_edited        = Column(Boolean, default=False, nullable=False)
    created_at       = Column(DateTime, server_default=func.now())
    updated_at       = Column(DateTime, server_default=func.now(), onupdate=func.now())

    __table_args__ = (
        UniqueConstraint("user_id", "target_type", "target_id", name="uq_review_per_target"),
    )

    author     = relationship("User", foreign_keys=[user_id])
    moderator  = relationship("User", foreign_keys=[reviewed_by])

File: lib/models/review.dart
enum ReviewTargetType {
  course,
  mentor,
  post,
  event;

  String get value => name;

  static ReviewTargetType fromValue(String v) =>
      ReviewTargetType.values.firstWhere((e) => e.value == v);
}

enum ReviewStatus {
  pending,
  approved,
  rejected;

  String get value => name;

  static ReviewStatus fromValue(String v) =>
      ReviewStatus.values.firstWhere((e) => e.value == v,
          orElse: () => ReviewStatus.pending);
}

class Review {
  final int id;
  final int userId;
  final String? authorName;
  final ReviewTargetType targetType;
  final int targetId;
  final int rating;
  final String? title;
  final String body;
  final ReviewStatus status;
  final String? rejectionReason;
  final bool isEdited;
  /// True when any moderation action has been taken.
  /// The actual moderator identity is never included in public/mediator responses.
  final bool moderated;
  /// Human-readable note from the API, e.g. "Reviewed by the moderation team."
  final String? moderationLabel;
  final DateTime? reviewedAt;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Review({
    required this.id,
    required this.userId,
    this.authorName,
    required this.targetType,
    required this.targetId,
    required this.rating,
    this.title,
    required this.body,
    required this.status,
    this.rejectionReason,
    required this.isEdited,
    this.moderated = false,
    this.moderationLabel,
    this.reviewedAt,
    this.createdAt,
    this.updatedAt,
  });

  factory Review.fromJson(Map<String, dynamic> j) => Review(
        id:               j['id'] as int,
        userId:           j['user_id'] as int,
        authorName:       j['author_name'] as String?,
        targetType:       ReviewTargetType.fromValue(j['target_type'] as String),
        targetId:         j['target_id'] as int,
        rating:           j['rating'] as int,
        title:            j['title'] as String?,
        body:             j['body'] as String,
        status:           ReviewStatus.fromValue(j['status'] as String? ?? 'pending'),
        rejectionReason:  j['rejection_reason'] as String?,
        isEdited:         j['is_edited'] as bool? ?? false,
        moderated:        j['moderated'] as bool? ?? false,
        moderationLabel:  j['moderation_label'] as String?,
        reviewedAt: j['reviewed_at'] != null
            ? DateTime.tryParse(j['reviewed_at'] as String)
            : null,
        createdAt: j['created_at'] != null
            ? DateTime.tryParse(j['created_at'] as String)
            : null,
        updatedAt: j['updated_at'] != null
            ? DateTime.tryParse(j['updated_at'] as String)
            : null,
      );
}

class ReviewSummary {
  final int count;
  final double? averageRating;

  const ReviewSummary({required this.count, this.averageRating});

  factory ReviewSummary.fromJson(Map<String, dynamic> j) => ReviewSummary(
        count:         j['count'] as int? ?? 0,
        averageRating: (j['average_rating'] as num?)?.toDouble(),
      );
}
